File: stitcher.js
const fs = require("fs");
const path = require("path");
const { execFileSync } = require("child_process");

const STREAM_ID = process.env.STREAM_ID;
const OUT_BASE = `/out/${STREAM_ID}`;
const OUT_HLS = path.join(OUT_BASE, "hls");
const OUT_DASH = path.join(OUT_BASE, "dash");

const AUDIO_SR = parseInt(process.env.AUDIO_SR || "48000", 10);
const AUDIO_CH = parseInt(process.env.AUDIO_CH || "2", 10);
const AAC_BITRATE = process.env.AAC_BITRATE || "128k";

const SEG_SECONDS = parseInt(process.env.SEG_SECONDS || "60", 10);
const HISTORY_HOURS = parseInt(process.env.HISTORY_HOURS || "12", 10);
const STITCH_INTERVAL_SEC = parseInt(
  process.env.STITCH_INTERVAL_SEC || "10",
  10
);

const WINDOW_MINUTES = HISTORY_HOURS * 60; // 720 for 12h
const SAFE_MARGIN_MINUTES = 1; // keep 1 minute margin for segment completion
const MIN_OK_SECONDS = SEG_SECONDS - 2; // accept "few seconds" loss, but ensure near-complete

function pad2(n) {
  return String(n).padStart(2, "0");
}

function utcMinuteKey(d) {
  return `${pad2(d.getUTCHours())}${pad2(d.getUTCMinutes())}`;
}

function keyToDateUTC(key) {
  // Key is now HHMM format, reconstruct date from current UTC date
  const now = new Date();
  const hh = parseInt(key.slice(0, 2), 10);
  const mm = parseInt(key.slice(2, 4), 10);
  return new Date(
    Date.UTC(
      now.getUTCFullYear(),
      now.getUTCMonth(),
      now.getUTCDate(),
      hh,
      mm,
      0
    )
  );
}

function run(cmd, args) {
  try {
    execFileSync(cmd, args, { stdio: ["ignore", "pipe", "pipe"] });
    return true;
  } catch (e) {
    // keep going, but log
    console.error(`[stitcher] command failed: ${cmd} ${args.join(" ")}`);
    if (e.stdout) console.error(String(e.stdout));
    if (e.stderr) console.error(String(e.stderr));
    return false;
  }
}

function ffprobeDurationSeconds(filePath) {
  try {
    const out = execFileSync(
      "ffprobe",
      [
        "-v",
        "error",
        "-show_entries",
        "format=duration",
        "-of",
        "default=noprint_wrappers=1:nokey=1",
        filePath,
      ],
      { stdio: ["ignore", "pipe", "pipe"] }
    )
      .toString()
      .trim();
    const dur = parseFloat(out);
    return Number.isFinite(dur) ? dur : null;
  } catch {
    return null;
  }
}

function ensureSilenceTs(tsPath) {
  // Write a 60s silent TS segment (AAC in TS)
  fs.mkdirSync(path.dirname(tsPath), { recursive: true });
  run("ffmpeg", [
    "-hide_banner",
    "-loglevel",
    "error",
    "-f",
    "lavfi",
    "-i",
    `anullsrc=r=${AUDIO_SR}:cl=${AUDIO_CH === 2 ? "stereo" : "mono"}`,
    "-t",
    String(SEG_SECONDS),
    "-c:a",
    "aac",
    "-b:a",
    AAC_BITRATE,
    "-ar",
    String(AUDIO_SR),
    "-ac",
    String(AUDIO_CH),
    "-f",
    "mpegts",
    tsPath,
  ]);
}

function ensureSilenceMp4(mp4Path) {
  // Write a 60s silent MP4 (AAC in MP4)
  fs.mkdirSync(path.dirname(mp4Path), { recursive: true });
  run("ffmpeg", [
    "-hide_banner",
    "-loglevel",
    "error",
    "-f",
    "lavfi",
    "-i",
    `anullsrc=r=${AUDIO_SR}:cl=${AUDIO_CH === 2 ? "stereo" : "mono"}`,
    "-t",
    String(SEG_SECONDS),
    "-c:a",
    "aac",
    "-b:a",
    AAC_BITRATE,
    "-ar",
    String(AUDIO_SR),
    "-ac",
    String(AUDIO_CH),
    "-movflags",
    "+faststart",
    mp4Path,
  ]);
}

function ensureMinuteFiles(minuteKey) {
  const tsPath = path.join(OUT_HLS, `${minuteKey}.ts`);
  const mp4Path = path.join(OUT_DASH, `${minuteKey}.mp4`);

  let needTsRecreate = false;
  let needMp4Recreate = false;

  // If exists but zero bytes or too short, replace with silence
  if (fs.existsSync(tsPath)) {
    const stats = fs.statSync(tsPath);
    if (stats.size === 0) {
      console.warn(
        `[stitcher] TS is zero bytes, replacing with silence: ${minuteKey}`
      );
      try {
        fs.unlinkSync(tsPath);
        needTsRecreate = true;
      } catch (e) {
        console.error(`[stitcher] Failed to delete zero-byte TS: ${e.message}`);
      }
    } else {
      const dur = ffprobeDurationSeconds(tsPath);
      if (dur !== null && dur < MIN_OK_SECONDS) {
        console.warn(
          `[stitcher] TS too short (${dur}s), replacing with silence: ${minuteKey}`
        );
        try {
          fs.unlinkSync(tsPath);
          needTsRecreate = true;
        } catch {}
      }
    }
  }
  if (fs.existsSync(mp4Path)) {
    const stats = fs.statSync(mp4Path);
    if (stats.size === 0) {
      console.warn(
        `[stitcher] MP4 is zero bytes, replacing with silence: ${minuteKey}`
      );
      try {
        fs.unlinkSync(mp4Path);
        needMp4Recreate = true;
      } catch (e) {
        console.error(
          `[stitcher] Failed to delete zero-byte MP4: ${e.message}`
        );
      }
    } else {
      const dur = ffprobeDurationSeconds(mp4Path);
      if (dur !== null && dur < MIN_OK_SECONDS) {
        console.warn(
          `[stitcher] MP4 too short (${dur}s), replacing with silence: ${minuteKey}`
        );
        try {
          fs.unlinkSync(mp4Path);
          needMp4Recreate = true;
        } catch {}
      }
    }
  }

  if (!fs.existsSync(tsPath) || needTsRecreate) {
    console.warn(
      `[stitcher] ${
        needTsRecreate ? "Recreating" : "Missing"
      } TS, creating silence: ${minuteKey}`
    );
    ensureSilenceTs(tsPath);
  }
  if (!fs.existsSync(mp4Path) || needMp4Recreate) {
    console.warn(
      `[stitcher] ${
        needMp4Recreate ? "Recreating" : "Missing"
      } MP4, creating silence: ${minuteKey}`
    );
    ensureSilenceMp4(mp4Path);
  }
}

function deleteOldFiles(keepFromKey) {
  // delete files older than keepFromKey- a small buffer
  const keepFromDate = keyToDateUTC(keepFromKey);
  const cutoffMs = keepFromDate.getTime() - 5 * 60 * 1000; // extra 5-min buffer

  for (const dir of [OUT_HLS, OUT_DASH]) {
    if (!fs.existsSync(dir)) continue;
    for (const name of fs.readdirSync(dir)) {
      const m = name.match(/^(\d{12})\.(ts|mp4)$/);
      if (!m) continue;
      const key = m[1];
      const dt = keyToDateUTC(key).getTime();
      if (dt < cutoffMs) {
        try {
          fs.unlinkSync(path.join(dir, name));
        } catch {}
      }
    }
  }
}

function writeHlsPlaylists(minuteKeys) {
  // HLS audio playlist: last 12h (720 segments), each 60s
  const targetDuration = SEG_SECONDS;
  const mediaSeq = 0; // we use absolute filenames; sequence is less important

  const lines = [];
  lines.push("#EXTM3U");
  lines.push("#EXT-X-VERSION:3");
  lines.push(`#EXT-X-TARGETDURATION:${targetDuration}`);
  lines.push(`#EXT-X-MEDIA-SEQUENCE:${mediaSeq}`);

  for (const key of minuteKeys) {
    const dt = keyToDateUTC(key);
    // program date time improves seeking semantics for many clients
    lines.push(`#EXT-X-PROGRAM-DATE-TIME:${dt.toISOString()}`);
    lines.push(`#EXTINF:${SEG_SECONDS.toFixed(3)},`);
    lines.push(`${key}.ts`);
  }

  const audioPath = path.join(OUT_HLS, "playlist.m3u8");
  fs.writeFileSync(audioPath, lines.join("\n") + "\n");

  // Master playlist
  const master = [];
  master.push("#EXTM3U");
  master.push("#EXT-X-VERSION:3");
  master.push(`#EXT-X-STREAM-INF:BANDWIDTH=160000,CODECS="mp4a.40.2"`);
  master.push("playlist.m3u8");

  fs.writeFileSync(path.join(OUT_HLS, "master.m3u8"), master.join("\n") + "\n");
}

function writeDashMpd(minuteKeys) {
  // Minimal dynamic MPD listing MP4-per-minute segments.
  // NOTE: This uses SegmentList. It’s restart-proof and simple for gap-fill.
  const now = new Date();
  const publishTime = now.toISOString();
  const timeShiftBufferDepth = `PT${HISTORY_HOURS}H`;
  const minUpdate = "PT30S";

  // Availability: start at earliest minute
  const ast = keyToDateUTC(minuteKeys[0]).toISOString();

  // Build SegmentURL list (relative paths)
  const segmentUrls = minuteKeys
    .map((k) => `      <SegmentURL media="${k}.mp4" />`)
    .join("\n");

  // Approx bandwidth hint (not critical)
  const bandwidth = 160000;

  const mpd = `<?xml version="1.0" encoding="UTF-8"?>
<MPD xmlns="urn:mpeg:dash:schema:mpd:2011"
     type="dynamic"
     availabilityStartTime="${ast}"
     publishTime="${publishTime}"
     minimumUpdatePeriod="${minUpdate}"
     timeShiftBufferDepth="${timeShiftBufferDepth}"
     profiles="urn:mpeg:dash:profile:isoff-live:2011">
  <Period id="p0" start="PT0S">
    <AdaptationSet id="a0" contentType="audio" mimeType="audio/mp4" segmentAlignment="true" lang="und">
      <Representation id="r0" bandwidth="${bandwidth}" codecs="mp4a.40.2" audioSamplingRate="${AUDIO_SR}">
        <AudioChannelConfiguration schemeIdUri="urn:mpeg:dash:23003:3:audio_channel_configuration:2011" value="${AUDIO_CH}"/>
        <SegmentList timescale="1" duration="${SEG_SECONDS}">
${segmentUrls}
        </SegmentList>
      </Representation>
    </AdaptationSet>
  </Period>
</MPD>
`;

  fs.writeFileSync(path.join(OUT_DASH, "manifest.mpd"), mpd);
}

function buildMinuteWindowKeys() {
  // We only finalize up to (now - SAFE_MARGIN_MINUTES).
  const now = new Date();
  const end = new Date(now.getTime());
  end.setUTCSeconds(0, 0);
  end.setUTCMinutes(end.getUTCMinutes() - SAFE_MARGIN_MINUTES);

  const keys = [];
  for (let i = WINDOW_MINUTES - 1; i >= 0; i--) {
    const d = new Date(end.getTime() - i * 60 * 1000);
    keys.push(utcMinuteKey(d));
  }
  return keys;
}

function tick() {
  try {
    fs.mkdirSync(OUT_HLS, { recursive: true });
    fs.mkdirSync(OUT_DASH, { recursive: true });

    const minuteKeys = buildMinuteWindowKeys();
    const keepFromKey = minuteKeys[0];

    // Ensure all minutes exist (real or silence)
    for (const key of minuteKeys) {
      ensureMinuteFiles(key);
    }

    // Remove old files
    deleteOldFiles(keepFromKey);

    // Write manifests
    writeHlsPlaylists(minuteKeys);
    writeDashMpd(minuteKeys);

    // Touch a heartbeat file for healthchecks if you want
    fs.writeFileSync(
      path.join(OUT_BASE, "stitcher_heartbeat.txt"),
      new Date().toISOString() + "\n"
    );
  } catch (e) {
    console.error("[stitcher] tick error:", e);
  }
}

console.log(
  `[stitcher] starting for ${STREAM_ID}, window=${WINDOW_MINUTES} minutes, seg=${SEG_SECONDS}s, UTC`
);
tick();
setInterval(tick, STITCH_INTERVAL_SEC * 1000);

File: stitcher/stitcher.py
import os, re, json, time
from datetime import datetime, timezone, timedelta

STREAM_ID = os.getenv("STREAM_ID", "stream1")
SEGMENT_DURATION = int(os.getenv("SEGMENT_DURATION", "60"))
TSBD = int(os.getenv("TIME_SHIFT_BUFFER_DEPTH", "43200"))
INTERVAL = int(os.getenv("STITCH_INTERVAL_SECONDS", "2"))

OUT_BASE = f"/output/{STREAM_ID}"

RUN_RE = re.compile(r"run_(\d+)\.json$")
SEG_RE = re.compile(r"seg_(\d+)_(\d+)\.m4s$")  # seg_<run>_<number>.m4s

def parse_utc(s: str) -> datetime:
    return datetime.fromisoformat(s.replace("Z", "+00:00")).astimezone(timezone.utc)

def list_runs():
    runs = []
    if not os.path.isdir(OUT_BASE):
        return runs
    for fn in os.listdir(OUT_BASE):
        m = RUN_RE.match(fn)
        if not m:
            continue
        path = os.path.join(OUT_BASE, fn)
        try:
            meta = json.load(open(path, "r", encoding="utf-8"))
            run_id = str(meta["run_id"])
            start_utc = parse_utc(meta["start_utc"])
            start_segment_number = int(meta.get("start_segment_number", 1))
            runs.append((run_id, start_utc, start_segment_number))
        except Exception:
            pass
    runs.sort(key=lambda x: int(x[0]))
    return runs

def list_segments():
    segs = []
    if not os.path.isdir(OUT_BASE):
        return segs
    for fn in os.listdir(OUT_BASE):
        m = SEG_RE.match(fn)
        if not m:
            continue
        run_id, num = m.group(1), int(m.group(2))
        segs.append((run_id, num, fn))
    segs.sort(key=lambda x: (int(x[0]), x[1]))
    return segs

def write_master():
    # Simple master referencing one audio rendition
    path = os.path.join(OUT_BASE, "playlist.m3u8")
    tmp = path + ".tmp"
    content = "\n".join([
        "#EXTM3U",
        "#EXT-X-VERSION:7",
        "#EXT-X-INDEPENDENT-SEGMENTS",
        f'#EXT-X-STREAM-INF:BANDWIDTH=96000,CODECS="mp4a.40.5"',
        "audio.m3u8",
        "",
    ])
    with open(tmp, "w", encoding="utf-8") as f:
        f.write(content)
    os.replace(tmp, path)

def write_hls_media(runs, segs):
    # Build a continuous media playlist with DISCONTINUITY across run boundaries
    # Keep only last TSBD seconds worth of segments (based on computed PDT)
    run_map = {rid: {"start": st, "start_num": sn} for rid, st, sn in runs}

    entries = []
    for run_id, num, fn in segs:
        if run_id not in run_map:
            continue
        start_time = run_map[run_id]["start"] + timedelta(seconds=(num - 1) * SEGMENT_DURATION)
        entries.append((start_time, run_id, num, fn))

    if not entries:
        return

    # retain last TSBD seconds (plus a small grace)
    newest_time = entries[-1][0]
    cutoff = newest_time - timedelta(seconds=TSBD)
    entries = [e for e in entries if e[0] >= cutoff]

    # HLS header
    targetduration = SEGMENT_DURATION
    media_seq = entries[0][2]  # not perfect across run ids, but ok for clients
    lines = [
        "#EXTM3U",
        "#EXT-X-VERSION:7",
        f"#EXT-X-TARGETDURATION:{targetduration}",
        f"#EXT-X-MEDIA-SEQUENCE:{media_seq}",
        "#EXT-X-PLAYLIST-TYPE:EVENT",
    ]

    prev_run = None
    prev_time = None

    # For fMP4 playlist, EXT-X-MAP must be set for each discontinuity/run
    for (t, run_id, num, fn) in entries:
        # gap detection
        if prev_time is not None:
            expected = prev_time + timedelta(seconds=SEGMENT_DURATION)
            if t > expected + timedelta(seconds=1):  # downtime gap
                lines.append("#EXT-X-DISCONTINUITY")
                prev_run = None  # force new EXT-X-MAP

        # run boundary = discontinuity + new init
        if prev_run is None or run_id != prev_run:
            if prev_run is not None:
                lines.append("#EXT-X-DISCONTINUITY")
            lines.append(f'#EXT-X-MAP:URI="init_{run_id}.mp4"')

        # program date time per segment (this is what keeps DVR “continuous”)
        lines.append(f"#EXT-X-PROGRAM-DATE-TIME:{t.isoformat().replace('+00:00','Z')}")
        lines.append(f"#EXTINF:{SEGMENT_DURATION:.3f},")
        lines.append(fn)

        prev_run = run_id
        prev_time = t

    path = os.path.join(OUT_BASE, "audio.m3u8")
    tmp = path + ".tmp"
    with open(tmp, "w", encoding="utf-8") as f:
        f.write("\n".join(lines) + "\n")
    os.replace(tmp, path)

def write_mpd(runs, segs):
    # Multi-period dynamic MPD. Each packager run becomes a Period.
    # Period start is derived from run start time relative to availabilityStartTime.
    if not runs:
        return

    ast = runs[0][1]  # availabilityStartTime = first run start
    now = datetime.now(timezone.utc)

    # which runs have segments at all?
    segs_by_run = {}
    for run_id, num, fn in segs:
        segs_by_run.setdefault(run_id, []).append(num)

    periods = []
    for run_id, start_utc, start_num in runs:
        nums = segs_by_run.get(run_id, [])
        if not nums:
            continue
        # Period@start in seconds since AST
        pstart = int((start_utc - ast).total_seconds())
        # Choose a conservative startNumber: smallest seen
        min_num = min(nums)
        # media name template
        periods.append((run_id, pstart, min_num))

    if not periods:
        return

    # Build MPD
    # NOTE: minimal MPD, good enough for most players, with gaps as separate Periods.
    mpd = []
    mpd.append('<?xml version="1.0" encoding="UTF-8"?>')
    mpd.append(
        f'<MPD xmlns="urn:mpeg:dash:schema:mpd:2011" '
        f'type="dynamic" '
        f'minimumUpdatePeriod="PT2S" '
        f'timeShiftBufferDepth="PT{TSBD}S" '
        f'availabilityStartTime="{ast.isoformat().replace("+00:00","Z")}">'
    )

    for run_id, pstart, start_number in periods:
        mpd.append(f'  <Period id="run_{run_id}" start="PT{pstart}S">')
        mpd.append('    <AdaptationSet id="1" contentType="audio" mimeType="audio/mp4" codecs="mp4a.40.5" lang="cs">')
        mpd.append('      <Representation id="audio" bandwidth="96000">')
        mpd.append(
            f'        <SegmentTemplate timescale="1" duration="{SEGMENT_DURATION}" '
            f'startNumber="{start_number}" '
            f'initialization="init_{run_id}.mp4" '
            f'media="seg_{run_id}_$Number$.m4s" />'
        )
        mpd.append('      </Representation>')
        mpd.append('    </AdaptationSet>')
        mpd.append('  </Period>')

    mpd.append('</MPD>')

    path = os.path.join(OUT_BASE, "manifest.mpd")
    tmp = path + ".tmp"
    with open(tmp, "w", encoding="utf-8") as f:
        f.write("\n".join(mpd) + "\n")
    os.replace(tmp, path)

def main():
    os.makedirs(OUT_BASE, exist_ok=True)
    print(f"[STITCHER] OUT_BASE={OUT_BASE} SEGMENT_DURATION={SEGMENT_DURATION} TSBD={TSBD} INTERVAL={INTERVAL}")

    while True:
        runs = list_runs()
        segs = list_segments()

        if runs and segs:
            write_master()
            write_hls_media(runs, segs)
            write_mpd(runs, segs)

        time.sleep(INTERVAL)

if __name__ == "__main__":
    main()